cloudinit/net/__init__.py

```diff
@@ -531,12 +535,14 @@ def find_candidate_nics_on_linux() -> List[str]:
     # 2. Remaining connected interfaces, naturally sorted.
     # 3. DEFAULT_PRIMARY_INTERFACE, if possibly connected.
     # 4. Remaining possibly connected interfaces, naturally sorted.
+    if not default_primary_interface:
+        default_primary_interface = DEFAULT_PRIMARY_INTERFACE
     sorted_interfaces = []
     for interfaces in [connected, possibly_connected]:
         interfaces = sorted(interfaces, key=natural_sort_key)
-        if DEFAULT_PRIMARY_INTERFACE in interfaces:
-            interfaces.remove(DEFAULT_PRIMARY_INTERFACE)
-            interfaces.insert(0, DEFAULT_PRIMARY_INTERFACE)
+        if default_primary_interface in interfaces:
```

Comment:
What this means now is that if any GCE multi-nic instance type doesn't have an `ens4` NIC in `interfaces` but happens to have `eth0` somwwhere in the sorted list, we will no longer insert `eth0` at index 0  because we will be searching exclusively for `ens4`.  

Did we want to limit the `ens4` default to certain instance types for google? If not, are certain we aren't inadvertently delaying some other google intsance types adversely by leaving `eth0` in a multi nic scenario unsorted as first?

Alternatively, is it worth an `else:` here  to attempt to insert this module's DEFAULT_PRIMARY_INTERFACE at index 0 if the parameter `default_primary_interface` wasn't found in interfaces.

Reply:
> Did we want to limit the ens4 default to certain instance types for google? If not, are certain we aren't inadvertently delaying some other google intsance types adversely by leaving eth0 in a multi nic scenario unsorted as first?

We cannot limit this to certain instance types because we have not yet the instance metadata and there is not enough DMI information to identify those machines. Additionally, they explicitly wanted to change the default globally (see attached SF ticket).

> Alternatively, is it worth an else: here to attempt to insert this module's DEFAULT_PRIMARY_INTERFACE at index 0 if the parameter default_primary_interface wasn't found in interfaces.

I went with this approach, as it is more backwards compatible.